lib/src/decorator.dart

```diff
@@ -0,0 +1,150 @@
+const String _browserClassPrefix = 'ua-';
+
+/// The string that will be prepended to the CSS class generated by [getPlatformClasses]
+/// for [_osClassName].
+const String _osClassPrefix = 'os-';
+
+/// The string that will be prepended to the CSS class generated by [getPlatformClasses]
+/// for [_osClassName].
+const String _versionRangeClassPrefix = 'lt-';
+
+/// The string that will be prepended to the CSS class generated by [getPlatformClasses]
+/// for [Feature.name] if [Feature.isSupported] is `false`.
+const String _featureSupportNegationClassPrefix = 'no-';
+
+/// The first major release after the current one.
+@visibleForTesting
+final int minVersionRange = browser.version.major + 1;
+
+/// The number of major releases above [minVersionRange]
+/// that [_getBrowserVersionClasses] should produce CSS classes for.
+@visibleForTesting
+const int maxVersionRange = 2;
+
+String _getBrowserVersionClasses() {
+  var majorVersion = browser.version.major;
+  var classes = [
+    _browserClassPrefix + _browserClassName + majorVersion.toString()
+  ];
+
+  for (var i = minVersionRange; i < minVersionRange + maxVersionRange; i++) {
+    classes.add(
+        '$_browserClassPrefix$_versionRangeClassPrefix$_browserClassName$i');
+  }
+
+  return _listToClassNameString(classes);
+}
+
+String _getFeatureSupportClasses(List<Feature> features) {
+  var classes = [];
+
+  for (var i = 0; i < features.length; i++) {
+    var feature = features[i];
+
+    classes.add(feature.isSupported
+        ? feature.name
+        : '${_featureSupportNegationClassPrefix}${feature.name}');
+  }
+
+  return _listToClassNameString(classes);
+}
+
+/// Utility fn that takes a list of [classes] and returns a space-separated string for
+/// use within the `class` attribute on the `<html>` element that gets injected by
+/// [decorateRootNodeWithPlatformClasses].
+String _listToClassNameString(List<String> classes) {
+  var classesStr = '';
+
+  for (var i = 0; i < classes.length; i++) {
+    if (classesStr.isEmpty) {
+      classesStr = classes[i];
+    } else {
+      classesStr += ' ${classes[i]}';
+    }
+  }
+
+  return classesStr;
+}
+
+/// Convert white-space within the given [name] to dashes, and convert it to lowercase
+/// for standardized CSS class formatting.
+@visibleForTesting
+String nameToClassName(String name) {
+  return name.replaceAll(' ', '-').toLowerCase();
+}
+
+/// Whether the given [rootNode] has already had it's CSS classes set via [decorateRootNodeWithPlatformClasses].
+bool nodeHasBeenDecorated(Element rootNode) =>
+    rootNode.classes.contains(_decorationCompleteClassName);
+
+/// Generates CSS classes based on the current [browser], [operatingSystem] and optionally,
+/// [features] that your app may need conditional styling for in addition to the
+/// [defaultFeatureCssClassDecorators] that will have CSS classes present by default.
+String getPlatformClasses({List<Feature> features, List<String> decorators}) {
+  decorators ??= [];
```

Comment:
If decorators is not null, it gets mutated below. What about rewriting the function like so?

```dart
String getPlatformClasses({List<Feature> features, List<String> existingClasses: const []}) {
  features ??= defaultFeatureCssClassDecorators;

  var classes = []
    ..addAll(existingClasses)
    ..add(_browserClassPrefix + _browserClassName)
    ..add(_getBrowserVersionClasses())
    ..add(_osClassPrefix + _osClassName)
    ..add(_getFeatureSupportClasses(features))
    ..add(_decorationCompleteClassName);

  return _listToClassNameString(classes);
}
```
This would not modify `features`/`decorators`, and also allows consumers to opt out of the default features by providing an empty list.